crypto/evp/evp_enc.c

```diff
@@ -930,9 +930,15 @@ int EVP_CIPHER_CTX_set_key_length(EVP_CIPHER_CTX *c, int keylen)
     params[0] = OSSL_PARAM_construct_int(OSSL_CIPHER_PARAM_KEYLEN, &keylen);
     ok = evp_do_ciph_ctx_setparams(c->cipher, c->provctx, params);
 
-    if (ok != EVP_CTRL_RET_UNSUPPORTED)
-        return ok;
+    if (ok == EVP_CTRL_RET_UNSUPPORTED)
+        goto legacy;
 
+    if (ok > 0 && EVP_CIPHER_CTX_key_length(c) != keylen) {
```

Comment:
That would work - but is this a case of smoking mirrors..
The assumption being that the get is handled correctly.
Imagine someone adds something that just needs to be handled by one special new cipher that only itself defines a get and set parameter.

Reply:
I think it can be a requirement that any NEW ciphers must handle the existing controls properly, as in signalling the right error and so on.

Reply:
It is not an error.. It will return ok if any parameters are missing.

Reply:
In the specific case of key length, I think we pretty much have to say getting a key length is mandatory to implement. Probably that also applies to iv length. (provider-cipher.pod should perhaps be updated with that info).

> Imagine someone adds something that just needs to be handled by one special new cipher that only itself defines a get and set parameter.

I'm not necessarily saying that this is a generic answer for the broader problem. Only that you don't need #9510/#9552 in order to solve the key length issue.